java/flight/src/main/java/org/apache/arrow/flight/ArrowMessage.java

```diff
@@ -387,6 +396,10 @@ public ArrowMessage parse(InputStream stream) {
 
   @Override
   public void close() throws Exception {
+    if (drained) {
```

Comment:
The problem is that we don't increase the buffer reference count in as input stream. Doing that would allow these to be closed out without worrying about whether the inputstream is depleted or not. I would avoid this drained pattern.

Reply:
Thanks for the suggestion - that works much better. I also then found a case where we weren't releasing buffers after writing dictionary buffers, so that should un-flake the tests of that.